[ENHANCEMENT] added temp table option for snowflake

talagluck opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Added option to use temp tables in snowflake in addition to the transient table option
- If the `"snowflake_transient_table"` key is present in batch_kwargs, we will use a transient table (which persists), otherwise we will default to a temp table
- To do this, we execute from a connection rather than an engine (as we do with sqlite and mssql)

Previous Design Review notes:
- After discussion with Eugene, we decided to continue to provide an option to use  transient tables so that this would not introduce a breaking change.

Thank you for submitting!

talagluck committed:
added default option to use temp tables in snowflake while maintaining ability to use transient tables

talagluck committed:
updated changelog

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
+                table_type = "TRANSIENT"
+            else:
+                table_type = "TEMPORARY"
+            logger.info("Creating temporary table %s" % table_name)
```
Just a suggestion to convert to f-strings when changing/adding an interpolated string (https://www.python.org/dev/peps/pep-0498/)
```suggestion
            logger.info(f"Creating temporary table {table_name}")
```

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
                 table_name = schema_name + "." + table_name
-            stmt = "CREATE OR REPLACE TRANSIENT TABLE {table_name} AS {custom_sql}".format(
-                table_name=table_name, custom_sql=custom_sql
+            stmt = "CREATE OR REPLACE {table_type} TABLE {table_name} AS {custom_sql}".format(
```
```suggestion
            stmt = f"CREATE OR REPLACE {table_type} TABLE {table_name} AS {custom_sql}"
```

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
-            stmt = "CREATE OR REPLACE TRANSIENT TABLE {table_name} AS {custom_sql}".format(
-                table_name=table_name, custom_sql=custom_sql
+            stmt = "CREATE OR REPLACE {table_type} TABLE {table_name} AS {custom_sql}".format(
+                table_type=table_type, table_name=table_name, custom_sql=custom_sql
```
```suggestion
                
```

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -1222,6 +1222,10 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
                 table_name=table_name, custom_sql=custom_sql
             )
+
+        # close_with_result = True
```
You might want to remove these commented out lines

talagluck replied:
Great catch - thanks!

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -1191,11 +1187,15 @@ def create_temporary_table(self, table_name, custom_sql, schema_name=None):
             )
         elif self.sql_engine_dialect.name.lower() == "snowflake":
-            logger.info("Creating transient table %s" % table_name)
+            if not self.generated_table_name:
```
Totally up to you on style, but you could do these 4 lines with a ternary assignment:
`table_type = "TEMPORARY" if self.generated_table_name else "TRANSIENT"`

talagluck replied:
Great suggestion - thanks!

anthonyburdi commented:
@talagluck Just a few suggestions! Take em or leave em.
@eugmandel Just to be clear I only did a code review, I did not consider the design.

anthonyburdi commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -461,9 +461,9 @@ def __init__(
             if engine.dialect.name.lower() == "mssql":
                 table_name = f"#{table_name}"
-            generated_table_name = table_name
+            self.generated_table_name = table_name
```
👍

talagluck committed:
quick fixes

eugmandel reviewed (commented):
@talagluck The code changes look good to me.

I don't see any changes in the CLI. Currently, it prompts for the transient table name. What do you think we should do? My thinking is to not require it in the CLI and have users configure it in the code because it will become a minority option.

Also, the how to guide for adding a Snowflake Datasource should reflect this change.

talagluck committed:
updated CLI and docs

talagluck commented:
Thanks, @eugmandel. Great point about the CLI - I updated the CLI to remove this prompt entirely.

I also updated the guide for adding a Snowflake Datasource.

talagluck committed:
Merge branch 'develop' into enhance/snowflake_temp_tables

eugmandel reviewed (approved):
@talagluck Thank you for getting the CLI and the guide changes in! I see Anthony has some code edit suggesting. Approving - your judgement which suggestions to accept before merging.

talagluck committed:
Merge branch 'develop' into enhance/snowflake_temp_tables
